### PR TITLE
HPCC-32971 Add removeAndSwapLast(n) to the array classes

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Untested, but pushed for discussion/review.

File: system/jlib/jarray.hpp [hunk 1]
```diff
@@ -281,6 +281,18 @@ class ArrayOf : public AllocatorOf<sizeof(MEMBER)>
...
     }
+    //Remove the element at pos and move the last element to pos - to avoid moving all elements
+    void removeAndSwapLast(aindex_t pos, bool nodestruct = false)
+    {
+        assertex(pos < SELF::used);
+        SELF::used --;
+        if (!nodestruct) SELF::destruct(pos);
+        if (pos != SELF::used)
+        {
+            MEMBER * head= (MEMBER *)SELF::_head;
```
Comment (jakesmith) [inline system/jlib/jarray.hpp:292]: trivial: space before =, probably c&p

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks correct to me.

**Comment by ghalliday** [comment]
@mckellyln rebased to target master + fixed spacing and missing include.  Please sanity check.

File: system/jlib/jarray.hpp [hunk 2]
```diff
@@ -281,6 +282,18 @@ class ArrayOf : public AllocatorOf<sizeof(MEMBER)>
         if (!nodestruct) SELF::destruct(pos);
         SELF::_move( pos, pos + 1, ( SELF::used - pos ) );
     }
+    //Remove the element at pos and move the last element to pos - to avoid moving all elements
+    void removeAndSwapLast(aindex_t pos, bool nodestruct = false)
+    {
+        assertex(pos < SELF::used);
```
Comment (mckellyln) [inline system/jlib/jarray.hpp:288]: Should we handle if this was called on the last element ?

**Review by mckellyln — APPROVED** [review]
Approved.

**Pull request merged**